src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala

```diff
@@ -0,0 +1,148 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.SparkSession
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.actions.Constants.States.{ACTIVE, OPTIMIZING}
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.index.IndexConstants.OPTIMIZE_MODES
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, OptimizeActionEvent}
+import com.microsoft.hyperspace.util.{HyperspaceConf, PathUtils}
+
+/**
+ * Optimize Action provides an optimize support for indexes where small index files
+ * can be merged into larger ones for better index performance.
+ *
+ * Algorithm outline:
+ * 1. Collect all the currently valid index files.
+ * 2. Split files into small and large, based on a threshold.
+ * 3. Bucketwise combine smaller files into 1 file per bucket.
+ * 4. Update index snapshot to remove small files and keep large files + newly created files.
+ *
+ * NOTE: This is an index-only operation. It does not look at the current state of
+ * the data at all. If the data was changed after index creation, optimize will NOT
+ * include the data changes.
+ *
+ * Available modes:
+ * `Quick` mode: This mode allows for fast optimization. Files smaller than a
+ * predefined threshold "spark.hyperspace.index.optimize.fileSizeThreshold" will be picked
+ * for compaction.
+ *
+ * `Full` mode: This allows for slow but complete optimization. ALL index files are
+ * picked for compaction.
+ *
+ * TODO: Optimize can be a no-op if there is only one optimizable file per bucket.
+ *  https://github.com/microsoft/hyperspace/issues/204.
+ *
+ * @param spark SparkSession.
+ * @param logManager IndexLogManager for index being refreshed.
+ * @param dataManager IndexDataManager for index being refreshed.
+ * @param mode Acceptable optimize modes are `quick` and `full`.
+ */
+class OptimizeAction(
+    spark: SparkSession,
+    final override protected val logManager: IndexLogManager,
+    dataManager: IndexDataManager,
+    mode: String)
+    extends CreateActionBase(dataManager)
+    with Action {
+  private lazy val previousLogEntry: LogEntry = {
+    logManager.getLog(baseId).getOrElse {
+      throw HyperspaceException("LogEntry must exist for optimize operation")
+    }
+  }
+
+  protected lazy val previousIndexLogEntry = previousLogEntry.asInstanceOf[IndexLogEntry]
```

Comment:
Any reason to separate these two out instead of:
```scala
  private lazy val previousIndexLogEntry: IndexLogEntry = {
    logManager.getLog(baseId) match {
      case Some(e: IndexLogEntry) => e
      case _ => throw HyperspaceException("IndexLogEntry must exist for optimize operation")
    }
  }
```
?

Reply:
thanks @imback82 , changed as suggested.